### DISCUSSION TITLE
variable substitution

### DISCUSSION THREAD

**Discussion by danielb2** [casey/just#1196] [Q&A] [answered]
```
SERVICES := "nfs postfix docker"

test:
  for i in {{SERVICES}}; do echo $i; done
  for i in $SERVICES; do echo $i; done
  for i in {{SERVICES}}; do echo {{i}}; done
  ```
  
  The first line passes, but second two fail. So what's the logic between `{{` and `$` here?

**Comment by casey** [comment, marked as answer]
In this case, `SERVICES` is a `just` variable, and `i` is a shell variable. You can export `just` variables to the shell with the export keyword, which will make the second line pass:

```
export SERVICES := "nfs postfix docker"

test:
  for i in {{SERVICES}}; do echo $i; done
  for i in $SERVICES; do echo $i; done
````

You can also import shell variables and use them in `just` expressions with the `env_var ` function:

```
x := env_var("FOO")

bar:
  echo {{x}}
```

**Reply by danielb2** [reply]
Thank you 🥰
